### PR TITLE
chore: Convert ErrorReporting to a Service to use DI. Add some tests (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
PAY-2113

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [x] Tests included.

### CODE REVIEW

File: packages/cli/src/error-reporting.ts [hunk 1]
```diff
@@ -1,106 +1,117 @@
...
-
-	setTag('server_type', Container.get(InstanceSettings).instanceType);
-
-	ErrorReporterProxy.init({
-		report: (error, options) => captureException(error, options),
-	});
-
-	initialized = true;
-};
+	beforeSend(event: ErrorEvent, { originalException }: EventHint) {
```
Comment (ivov) [inline packages/cli/src/error-reporting.ts:89]:
```suggestion
	private beforeSend(event: ErrorEvent, { originalException }: EventHint) {
```

File: packages/cli/src/error-reporting.ts [hunk 2]
```diff
@@ -1,106 +1,117 @@
...
-	if (!dsn) {
-		initialized = true;
-		return;
+import { Service } from 'typedi';
+
+@Service()
+export class ErrorReporting {
+	private initialized = false;
+
+	private seenErrors = new Set<string>();
```
Comment (ivov) [inline packages/cli/src/error-reporting.ts:15]:
```suggestion
  /** Hashes of error stack traces, to deduplicate error reports. */
	private readonly seenErrors = new Set<string>();
```

**Review by ivov — COMMENTED** [review]
🙏🏻

File: packages/cli/src/error-reporting.ts (hunk 1 above)
Comment (netroy) [reply packages/cli/src/error-reporting.ts]: we have tests for this. can't make it private without creating a typing error.

File: packages/cli/src/error-reporting.ts [hunk 3]
```diff
@@ -1,106 +1,117 @@
...
+
+	constructor(
+		private readonly globalConfig: GlobalConfig,
+		private readonly instanceSettings: InstanceSettings,
+	) {}
+
+	async init() {
+		if (this.initialized) return;
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/error-reporting.ts]: What makes you doubt? My reasoning is that initializing this twice means a developer error, so we could have an assertion error enforcing this. From what I see `init` is called only from `BaseCommand`.

File: packages/cli/src/error-reporting.ts (hunk 1 above)
Comment (ivov) [reply packages/cli/src/error-reporting.ts]: Never seen a good solution for this. Often it's useful to test private methods but that means TS exceptions :/

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.73.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.73.0)

**Pull request merged**